nntrainer/include/optimizer.h

```diff
@@ -0,0 +1,177 @@
+/**
+ * Copyright (C) 2020 Samsung Electronics Co., Ltd. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @file	optimizer.h
+ * @date	08 April 2020
+ * @brief	This is Optimizer classes of Neural Network
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ */
+#ifndef __OPTIMIZER_H__
+#define __OPTIMIZER_H__
+
+#include <iostream>
+#include "tensor.h"
+
+/**
+ * @brief     Enumeration of Optimizer
+ *            0. SGD
+ *            1. ADAM
+ *            2. Unknown
+ */
+enum class OptType
+{
+  sgd = 0,
+  adam = 1,
+  unkonwn = 2
+};
+
+/**
+ * @brief     Enumeration of Weight Decay type
+ *            0. L2Norm
+ *            1. Regression
+ *            2. Unknown
+ */
+enum class WeightDecayType
+{
+  l2norm = 0,
+  regression = 1,
+  unknown = 2
+};
+
+/**
+ * @brief     type for the Weight Decay hyper-parameter
+ */
+typedef struct {
+  WeightDecayType type;
+  float lambda;
+} weight_decay_param;
+
+/**
+ * @brief     type for the Optimizor to save hyper-parameter
+ */
+typedef struct {
+  float learning_rate;
+  double beta1;
+  double beta2;
+  double epsilon;
+  float decay_rate;
+  float decay_steps;
+  weight_decay_param weight_decay;
+} OptParam;
```

Comment:
Q: "weight_decay_param" is follwed by the snake naming convention while OptPram is followed by the camel naming convention. Is there any difference between two typedefed structures?

Reply:
I will change it later for all codes. But the rule is as below.
- Type Names : Pascal Naming rule ( OptParam )
- Variable Names : snake Naming rule ( weight_decay )
- Function Name : Camel Naming rule (setFunction())
- Enumerator : Upper Capital with underscore.
if you have better suggestion, then please let me know.